lib/luanetfilter.c

```diff
@@ -0,0 +1,182 @@
+
+#include <lua.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+#include "luanetfilter.h"
+#include "luadata.h"
+
+typedef struct luanetfilter_s {
+	lunatik_object_t *runtime;
+	lunatik_object_t *skb;
+	struct nf_hook_ops nfops;
+} luanetfilter_t;
+
+static void luanetfilter_release(void *private);
+
+static int luanetfilter_hook_cb(lua_State *L, luanetfilter_t *luanf, struct sk_buff *skb)
+{
+	int ret = -1;
+	if (lunatik_getregistry(L, luanf) != LUA_TTABLE) {
+		pr_err("lunatik hook: could not find ops table\n");
+		goto err;
+	}
+
+	if (lua_getfield(L, -1, "hook") != LUA_TFUNCTION) {
+		pr_err("luanetfilter hook: operation not defined");
+		goto err;
+	}
+
+	if (lunatik_getregistry(L, luanf->skb) != LUA_TUSERDATA) {
+		pr_err("luanetfilter hook: could not find skb");
+		goto err;
+	}
+	lunatik_object_t *data = (lunatik_object_t *)lunatik_toobject(L, -1);
+	if (unlikely(data == NULL || skb_linearize(skb) != 0)) {
+		pr_err("could not get skb\n");
+		return -1;
+	}
+	luadata_reset(data, skb->data, skb_headlen(skb), LUADATA_OPT_NONE);
+
+	if (lua_pcall(L, 1, 1, 0) != LUA_OK) {
+		pr_err("luanetfilter hook: pcall error %s\n", lua_tostring(L, -1));
+		goto err;
+	}
+	ret = lua_tointeger(L, -1);
+err:
+	return ret;
+}
+
+static inline unsigned int luanetfilter_docall(luanetfilter_t *luanf, struct sk_buff *skb)
+{
+	int ret;
+	if (!luanf || !luanf->runtime) {
+		pr_err("lunatik: netfilter runtime not found\n");
+		return NF_ACCEPT;
+	}
+
+	lunatik_run(luanf->runtime, luanetfilter_hook_cb, ret, luanf, skb);
+	return (ret < 0 || ret > NF_MAX_VERDICT) ? NF_ACCEPT : ret;
+}
+
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 4, 0))
+static unsigned int luanetfilter_hook(void *priv, struct sk_buff *skb, const struct nf_hook_state *state)
+{
+	luanetfilter_t *luanf = (luanetfilter_t *)priv;
+	return luanetfilter_docall(luanf, skb);
+}
+#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 1, 0))
+static unsigned int luanetfilter_hook(const struct nf_hook_ops *ops, struct sk_buff *skb, const struct nf_hook_state *state)
+{
+	luanetfilter_t *luanf = (luanetfilter_t *)ops->priv;
+	return luanetfilter_docall(luanf, skb);
+}
+#else
+static unsigned int luanetfilter_hook(const struct nf_hook_ops *ops, struct sk_buff *skb, const struct net_device *in, const struct net_device *out, int (*okfn)(struct sk_buff *))
+{
+	luanetfilter_t *luanf = (luanetfilter_t *)ops->priv;
+	return luanetfilter_docall(luanf, skb);
+}
+#endif
+
+static const luaL_Reg luanetfilter_mt[] = {
+	{"__gc", lunatik_deleteobject},
+	{NULL, NULL}
+};
+
+static const lunatik_class_t luanetfilter_class = {
+	.name = "netfilter",
+	.methods = luanetfilter_mt,
+	.release = luanetfilter_release,
+	.sleep = false,
+};
+
+#define luanetfilter_setinteger(L, idx, hook, field) 		\
+do {								\
+	lunatik_checkfield(L, idx, #field, LUA_TNUMBER);	\
+	hook->field = lua_tointeger(L, -1);			\
+	lua_pop(L, 1);						\
+} while (0)
+
+static inline void luanetfilter_newbuffer(lua_State *L, int idx, luanetfilter_t *luanf)
```

Comment:
There are three common functions here :
1. `lua{xtable/netfilter}_setinteger(L, idx, hook, field)` : used to fetch the integer from a table at index `idx` and set it to `hook->field`.

2. `lua{xtable/netfilter}_setstring(L, idx, hook, field)` : used to fetch the string from a table at index `idx` and set it to `hook->field`.

3. `lua{xtable/netfilter}_newbuffer(L, idx, luaxtable *)` : used to create a `luadata` object pointing to `NULL` with no options and this is set to xtable/netfilter->skb;

I think we can move these to lunatik headers, can you suggest a good prefix name for this (`lunatik_xyz_setinteger` ??)

Reply:
I think let's not move (3) to lunatik, since it depends on luadata, I think it will be better to keep it seperate from such dependency, what do you suggest ?

Reply:
I would just move them to `luanetfilter.h` as inline functions and use them on `luaxtables.c`.. what do you think?